signac/core/synced_collections/synced_attr_dict.py

```diff
@@ -212,36 +215,43 @@ def reset(self, data=None):
             )
 
     def keys(self):  # noqa: D102
-        self._load()
+        with self._thread_lock():
+            self._load()
         return self._data.keys()
 
     def values(self):  # noqa: D102
-        self._load()
+        with self._thread_lock():
+            self._load()
         return self._to_base().values()
 
     def items(self):  # noqa: D102
-        self._load()
+        with self._thread_lock():
+            self._load()
         return self._to_base().items()
 
     def get(self, key, default=None):  # noqa: D102
-        self._load()
+        with self._thread_lock():
+            self._load()
         return self._data.get(key, default)
 
     def pop(self, key, default=None):  # noqa: D102
-        self._load()
-        ret = self._data.pop(key, default)
-        self._save()
+        with self._thread_lock():
+            self._load()
+            ret = self._data.pop(key, default)
+            self._save()
         return ret
 
     def popitem(self):  # noqa: D102
-        self._load()
-        ret = self._data.popitem()
-        self._save()
+        with self._thread_lock():
+            self._load()
+            ret = self._data.popitem()
+            self._save()
         return ret
 
     def clear(self):  # noqa: D102
         self._data = {}
-        self._save()
+        with self._thread_lock():
+            self._save()
```

Comment:
Do we want to overwrite the internal contents of `self._data` before the thread lock is acquired? I'm not sure.

Reply:
Doing so should be safe. In theory it's better for performance, since the Python interpreter could do that assignment concurrently with I/O on other threads. In practice it'll never be noticeable, I just aimed to lock the smallest portions of the code necessary for thread safety to minimize the impact everywhere and this is what happens here.